### PR TITLE
Add word count label to editor

### LINKED ISSUE #10176: Editor: Word count
A number of users have requested a word/character count in the editor.

Currently in Calypso web this appears in the bottom right corner of the editor:

Previous issue here: https://github.com/wordpress-mobile/WordPress-Editor-iOS/issues/606

### PR DESCRIPTION
**Pull request by yaelirub**
Added a word count function to UITextView.
  When editing/ creating post the word count label appears on HTML or Rich Text style when then number of words greater that 0

Fixes #10176 

To test:
* New Post *
- Create a new post 
- Type something in title
- Type something in text view. Notice that the word count label appears after the first word  
- Continue typing and notice the word count increases with every new word
- remove all content words and notice the label disappears when there are no words. 

* Editing post *
- Edit an existing post
- Notice that the word count label appears if the post had content
- Continue editing and notice the word count increases with every new word

* Same functionality should work for HTML and for Rich text.

word_count

### CODE REVIEW

**Comment by koke** [comment]
I know there are some languages where this doesn’t make sense since they don’t use spaces between words. I checked and calypso seems to disable the word count for those languages:
https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-word-count/index.jsx

**Comment by yaelirub** [comment]
TIL! Thanks @koke! I'll add character count functionality for those languages then 🙏

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 1]
```diff
@@ -77,6 +77,19 @@ class AztecPostViewController: UIViewController, PostEditor {
...
     }()
+    
+    fileprivate(set) lazy var wordCountLabel: UILabel = {
+        let label = UILabel()
+        label.textColor = Colors.wordCount
+        label.font = Fonts.regular
+        label.isUserInteractionEnabled = false
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.numberOfLines = 1
+        label.textAlignment = .right
```
[the remark this answers is not in the document]
Comment (jklausa) [reply WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: I don't think we should disable this check here, it makes sense to me for the label to display on the other side of the screen for the RTL languages.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift [hunk 2]
```diff
@@ -78,6 +78,42 @@ class AztecPostViewController: UIViewController, PostEditor {
...
+        return label
+    }()
+
+    /// If the language the user chose in settings is character based
+    /// The lable will show the number of characters.
+    /// If the language is not supported the label will stay hidden
+    /// otherwise, show word count.
+    ///
+
+    fileprivate(set) lazy var isCharacterBasedLanguge: Bool? = {
```
Comment (jklausa) [inline WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: Two things here: I think `Bool?` is rarely a useful return type for functions/properties. It's makes the code at call-site near impossible to understand without reading the implementation, but thankfully There's A Better Way!

I really like modeling problems like this with simple, throwaway Swift enums, something like this:
(pardon the naming, it's just an example, but my jetlagged brain can't come up with anything better at this hour, hopefully it gets the point across):
```swift
enum WordLabelDisplayMode {
    case wordBased
    case characterBased
    case hidden
}
```

It's much more descriptive and makes reading the caller site a breeze :)

But more importantly: I'm not sure whether this is the _right thing to do here_. I know nothing about japanese/chinese/korean/thai languages, so I'm not sure whether "character count" is a useful replacement metric for those languages. 

I'd rather us hide it entirely (which is what Web seems to be doing), than do something that's culturally insensitive or just doesn't make sense. Perhaps we could get some polyglots to chime in here whether that makes sense?

**Comment by jklausa** [comment]
One last thought: I think this issue might look deceptively simple, but given how _complicated_ our Text views are (and, just, like, Text in general!) it might be too much to take on right now: maybe a good middle-ground for this would be for the word count to be slightly more hidden? One place that comes to mind is the "Post Settings" screen, which, while not ideal, still lets users see how long the post is, and lets us sidestep a bunch of issues with UI here.

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (yaelirub) [reply WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: ooh right! I mean left! ;)

File: WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift (hunk 2 above)
Comment (yaelirub) [reply WordPress/Classes/ViewRelated/Aztec/ViewControllers/AztecPostViewController.swift]: 🙏 Thank you for this comment! Turning this to an enum looks much better! 
For the character counting. I saw a comment on the Web project saying character count should be added. But I know consistency is really important so Im gonna remove this and just not show the count in those cases.

**Comment by yaelirub** [comment]
Hey @jklausa , updated per your suggestions. 
word_counting

The issue with the line covering the word count label still exists though. I added a UIVisualEffectView with a blur but it still doesn't look right, so wondering if we could get input from design on this, or maybe do what you suggested and add this as optional functionality in the post settings?

**Comment by jklausa** [comment]
^- this should summon our designers :)

**Comment by yaelirub** [comment]
Closing this PR and will open a new one that reflects design input (show word and character count in action sheet).

**Pull request closed without merging**